AEFeinstein/Swadge-IDF-5.0
# Quick Settings Menu

### Description

This is an implementation of the Quick Settings Menu described in #28. It includes a new mode for the quick settings menu, as well as a new menu renderer for the quick setting. Some common menu functionality was also broken out into a separate file to keep things clean and reduce the amount of coupling with the `menuItem_t` struct.

It's not styled at all other than having rounded corners, but it has icons for on/off for LEDs, BGM, and SFX, and more settings and icons can be easily added if needed (TFT backlight?).

Some things are a bit funky and could probably be cleaned up a bit with some API additions:
* Rotating the direction of the menu
* Changing the value of a menu item / its selected option involves a lot of poking at the struct
* The adding extra info to menu items thing that the renderer does is a bit clunky, maybe a data pointer in the menu item?
* Having a separate mute/disable setting for sfx/bgm and leds would make things cleaner and a bit more user friendly

A screenshot of the quick settings menu over top of the pong mode:
![the quick settings menu draws over the lazy pong](https://github.com/AEFeinstein/Swadge-IDF-5.0/assets/666431/09317d6b-24ef-429e-8065-8ecf8e1dab26)




### Test Instructions

To enter the quick settings menu, hold the START button for .5 seconds from any mode other than the main menu. To exit the quick settings menu, hold the START button for .5 seconds again. A bar should be drawn across the top of the screen to indicate the timer progress. You should not be able to exit to the main menu until the quick settings menu is closed. 

To change between settings, press left and right. To turn a setting on, press up. To turn a setting off, press down. To toggle a setting, press A. If the setting was on (i.e. not set to its min value) at the time the quick settings menu was opened, that value will be used as the "on" value. If the setting was off at the time the quick settings menu was opened, the setting's maximum value will be used as the "on" value.



### Ticket Links

Closes #28 



### Readiness Checklist
